include/tscpp/util/Bravo.h

```diff
@@ -0,0 +1,386 @@
+
+  ~shared_lock()
+  {
+    if (_owns) {
+      _mutex->unlock_shared(_token);
+    }
+  };
+
+  ////
+  // Not Copyable
+  //
+  shared_lock(shared_lock const &)            = delete;
+  shared_lock &operator=(shared_lock const &) = delete;
+
+  ////
+  // Moveable
+  //
+  shared_lock(shared_lock &&s) : _mutex(s._mutex), _token(s._token), _owns(s._owns)
+  {
+    s._mutex = nullptr;
+    s._token = 0;
+    s._owns  = false;
+  };
+
+  shared_lock &
+  operator=(shared_lock &&s)
+  {
+    if (_owns) {
+      _mutex->unlock_shared(_token);
+    }
+    _mutex = s._mutex;
+    _token = s._token;
+    _owns  = s._owns;
+
+    s._mutex = nullptr;
+    s._token = 0;
+    s._owns  = false;
+  };
+
+  ////
+  // Shared locking
+  //
+  void
+  lock()
+  {
+    _mutex->lock_shared(_token);
+    _owns = true;
+  }
+
+  bool
+  try_lock()
+  {
+    _owns = _mutex->try_lock_shared(_token);
+    return _owns;
+  }
+
+  // not implemented yet
+  bool try_lock_for()   = delete;
+  bool try_lock_until() = delete;
+
+  void
+  unlock()
+  {
+    _mutex->unlock_shared(_token);
+    _owns = false;
+  }
+
+  ////
+  // Modifiers
+  //
+  void
+  swap(shared_lock &s)
+  {
+    std::swap(_mutex, s._mutex);
+    std::swap(_token, s._token);
+    std::swap(_owns, s._owns);
+  }
+
+  mutex_type *
+  release()
+  {
+    mutex_type *m = _mutex;
+    _mutex        = nullptr;
+    _token        = 0;
+    _owns         = false;
+    return m;
+  }
+
+  ////
+  // Observers
+  //
+  mutex_type *
+  mutex()
+  {
+    return _mutex;
+  }
+
+  Token
+  token()
+  {
+    return _token;
+  }
+
+  bool
+  owns_lock()
+  {
+    return _owns;
+  }
+
+private:
+  mutex_type *_mutex = nullptr;
+  Token _token       = 0;
+  bool _owns         = false;
+};
+
+/**
+   ts::bravo::shared_mutex
+
+   You can use std::lock_guard for writers but, you can't use std::shared_lock for readers to handle ts::bravo::Token.
+   Use ts::bravo::shared_lock for readers.
+
+   SLOT_SIZE needs to be larget than number of threads for lock_shared to go fast-path.
+ */
+template <typename T = std::shared_mutex, size_t SLOT_SIZE = 4096, int SLOWDOWN_GUARD = 7> class shared_mutex_impl
+{
+public:
+  shared_mutex_impl()  = default;
+  ~shared_mutex_impl() = default;
+
+  ////
+  // No copying or moving.
+  //
+  shared_mutex_impl(shared_mutex_impl const &)            = delete;
+  shared_mutex_impl &operator=(shared_mutex_impl const &) = delete;
+
+  shared_mutex_impl(shared_mutex_impl &&)            = delete;
+  shared_mutex_impl &operator=(shared_mutex_impl &&) = delete;
+
+  ////
+  // Exclusive locking
+  //
+  void
+  lock()
+  {
+    _mutex.underlying.lock();
+    _revoke();
+  }
+
+  bool
+  try_lock()
+  {
+    bool r = _mutex.underlying.try_lock();
+    if (!r) {
+      return false;
+    }
+
+    _revoke();
+
+    return true;
+  }
+
+  void
+  unlock()
+  {
+    _mutex.underlying.unlock();
+  }
+
+  ////
+  // Shared locking
+  //
+  void
+  lock_shared(Token &token)
```

Comment:
Why can't  lock_shared() just return the token.  Probably not faster after optimization, but seems clearer to me.